Add new autoproviders for geth --dev and Infura #797

bartleyg opened the pull request:
### What was wrong?

Added new autoproviders for ```geth --dev``` and Infura related to Issue #797 

### How was it fixed?

Users can now:
- [x] ```from web3.auto.gethdev import w3``` to connect to the default ```geth --dev``` location
- [x] ```from web3.auto.infura import w3``` to connect to Infura's Mainnet remote node

Also when users:
- [x] ```from web3.auto import w3``` the new ```geth --dev``` and Infura provider connections were added to the end of the default list of provider connections to try. The order the AutoProvider attempts to connect to is now:
1.  provider defined in environment variable
2.  IPCProvider
3.  HTTPProvider
4.  WebsocketProvider
5.  geth --dev IPCProvider
6.  Infura Mainnet HTTPProvider

This now means using ```from web3.auto import w3``` will always connect successfully, assuming either a local provider is found at the default locations or Infura's Mainnet remote node is operational.

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://i.imgur.com/NyvWIEK.png)

carver commented on web3/providers/ipc.py:
```diff
@@ -136,6 +136,50 @@ def get_default_ipc_path(testnet=False):
+            ".",
+            "pipe",
+            "geth.ipc"
+        )
```
Geth opens the same IPC endpoint for dev and production on windows?

bartleyg replied:
On Windows Server 2016 I see the same IPC endpoint (geth.ipc) url populate in the console of ```geth --dev```. I have not seen the jsonrpc.ipc endpoint on dev or production windows, so I left it in there like ```get_default_ipc_path()``` had. I currently don't have access to another Windows variant to check endpoint paths.

pipermerriam commented on web3/auto/infura.py:
```diff
@@ -0,0 +1,9 @@
+    INFURA_MAINNET_HTTP_URL,
+)
+
+w3 = Web3(HTTPProvider(INFURA_MAINNET_HTTP_URL))
```
Maybe we can allow that to be specified by environment variable and either raise a warning if it isn't present or error if we want to be strict.

carver replied:
I like it. I briefly looked for a standard name and didn't find any. One blog post used `INFURA_API_KEY`, which seems reasonable to me.

bartleyg commented on web3/auto/gethdev.py:
```diff
@@ -0,0 +1,9 @@
+    get_dev_ipc_path,
+)
+
+w3 = Web3(IPCProvider(get_dev_ipc_path()))
```
Ok I'll add the middleware.

bartleyg commented:
All great concerns so I removed gethdev and infura from the AutoProvider in the latest commit.

I added the ```geth_poa_middleware``` to web3/auto/gethdev.py. I didn't use ```load_gethdev_ipc_provider()``` as it was only needed for the AutoProvider logic.

 Also I moved the infura url constant to web3/auto/infura.py since it's not in the AutoProvider anymore. Some related tests were also moved around.

I left web3/auto/infura.py in for now until the decision is made to either remove it or add something like an API key environment variable requirement to run it.

carver commented:
Great, just the api key and warning, and we should be good to go.

Let's look for the environment variable `INFURA_API_KEY`. If it's available, use it. If not, connect anyway, but run a `logging.warning()` that includes the name of the environment variable and ideally provides a link to where you should go to request a new key.

carver commented:
Actually, I am putting out a release today, and would love to see this in it. Unless you can wrap this in the next few hours @bartleyg - I'll just add the last tweak on top and close it. Ping me either way, if you're around.

bartleyg commented:
@carver please be my guest and tweak it out for the release 👍

bartleyg commented:
Awesome, thanks for the assist @carver !
